hibernate-core/src/main/java/org/hibernate/event/internal/DefaultFlushEntityEventListener.java

```diff
@@ -332,15 +332,20 @@ protected boolean handleInterception(FlushEntityEvent event) {
 		final boolean intercepted = invokeInterceptor( session, entity, entry, values, persister );
 
 		//now we might need to recalculate the dirtyProperties array
-		if ( intercepted && event.isDirtyCheckPossible() && !event.isDirtyCheckHandledByInterceptor() ) {
-			int[] dirtyProperties;
-			if ( event.hasDatabaseSnapshot() ) {
-				dirtyProperties = persister.findModified( event.getDatabaseSnapshot(), values, entity, session );
+		if ( intercepted && event.isDirtyCheckPossible() ) {
+			if ( !event.isDirtyCheckHandledByInterceptor() ) {
+				int[] dirtyProperties;
+				if ( event.hasDatabaseSnapshot() ) {
+					dirtyProperties = persister.findModified( event.getDatabaseSnapshot(), values, entity, session );
+				}
+				else {
+					dirtyProperties = persister.findDirty( values, entry.getLoadedState(), entity, session );
+				}
+				event.setDirtyProperties( dirtyProperties );
 			}
 			else {
-				dirtyProperties = persister.findDirty( values, entry.getLoadedState(), entity, session );
+				dirtyCheck( event );
 			}
-			event.setDirtyProperties( dirtyProperties );
```

Comment:
This set of changes seems to fix the bug by ensuring a second dirty check is performed after a `@PreUpdate`, and even fixes an additional bug: it makes sure that, after a `FlushEntityEvent` is intercepted, bytecode enhancement or the custom dirtiness strategy (if any) is used to compute dirtiness instead of `persister.findModified` or `persister.findDirty` like we used to do.
However it's partly because another bug offsets problems in this implementation, and this "additional bug" might very well resurface later... So I'd like to propose going one step further.

But before that, a little explaining is necessary:

As Guillaume pointed out to me, the flag `isDirtyCheckHandledByInterceptor` is set to true when using bytecode enhancement or a custom dirtiness strategy: see in `dirtyCheck` how the `interceptorHandledDirtyCheck` variable is set to `true` if the properties are non-null after the bytecode enhancement/custom strategy block.

As a result, the new implementation of `handleInterception` in this patch will call `dirtyCheck` not only when an interceptor handled the previous dirty check, but also when bytecode enhancement or a custom dirty strategy handled the dirty check. Which is good... but it's because of a bug!

I pushed a few commits to this PR to demonstrate, and fix, the issue.

Commit 1: I fixed the `isDirtyCheckHandledByInterceptor`. The bytecode enhancement tests still pass, but only because, after the `@PreUpdate` is applied, we default to calling `persister.findDirty`, and it happens to work just as well as bytecode enhancement. Not sure it's a good idea to do that, though...
Commit 2: I added a test checking that, when using a custom dirtiness strategy, the strategy is used to compute dirtiness both before and after the `@PreUpdate`. The test fails.
Commit 3: I changed the `handleInterception` method to call the `dirtyCheck` method directly when dirty checking is necessary, instead of trying to inline an incomplete version of `dirtyCheck` inside `handleInterception`.
Commit 4: just a small refactorying clarifying a variable name.

Commit 3 is the core of the proposal: since we want to perform dirty checking a second time, let's just call `dirtyCheck`, which after all is here just for that. I think the behavior will be more correct, and from what I can see, there is little risk of experiencing a performance hit for the two cases that were already working:
 * the load state (used in `persister.findDirty`) is available as part of the entity entry and doesn't require any database call.
 * the database state (used in `persister.findModified`) might be loaded from the database, we're talking about the second pass (after the `@PreUpdate`), so if this method is being used, then it was already used during the first pass (before the `@PreUpdate`), and `org.hibernate.engine.internal.StatefulPersistenceContext#getDatabaseSnapshot` relies on a cache. So we won't be fetching data from the database on the second pass.

@gsmet , @vladmihalcea , WDYT?